### PR TITLE
Incremental month building

### LINKED ISSUE #6: link.sh defaults to linking everything at once in the background
I just noticed that the link script invokes everything in parallel, as background jobs:

https://github.com/elfshaker/elfshaker/blob/f69cb7bc70e7f7579669d5e0f873632d4406c453/contrib/manyclangs-build#L44

This is fine if you have lots of resources, but in a resource constrained environment such as a laptop I can't imagine this works very well. Additionally, if you accidentally discover this, you can't ctrl-c these processes since they are run in the background, so they don't receive ctrl-c.

One possible route is to `export -f $NAME` all of the functions, then do `| xargs -P$NCPU -n1 bash -c`, with an appropriate value for NCPU. This would mean that at least we wouldn't start more links than there are CPUs, which might be a better start, and the processes wouldn't run in the background and therefore would respond to ctrl-c.

### LINKED ISSUE #11: Clang relies on __PRETTY_FUNCTION__ and asserts
IIRC, we observed that use of `__PRETTY_FUNCTION__` resulted in binaries which depended on the full path of the sources. In order to make the binaries reproducible, we have a workaround:
https://github.com/elfshaker/elfshaker/blob/f69cb7bc70e7f7579669d5e0f873632d4406c453/contrib/manyclangs-cmake#L13-L14

Unfortunately the resulting clang binary is completely broken and refuses to compile anything:

https://llvm.org/doxygen/TypeName_8h_source.html

It asserts on `assert(!Name.empty() && "Unable to find the template parameter!");`. I suspect that __func__ doesn't contain the template type parameter information in it.

We'll need to find a workaround for this or stop overriding `__PRETTY_FUNCTION__` and therefore have binaries which can only be reproduced if built at a specific absolute path.

### PR DESCRIPTION
**Pull request by peterwaller-arm**
This is significant rework of the manyclangs build system.

Fixes #11.
Fixes #6.

# TODO
- [x] Address comments.
- [x] link.sh symlinks don't currently work.

My goals have included:

* UX Simplicity.
  * To reproduce a build the command `manyclangs-build [version string]`,
    from a build directory nested under llvm-project is sufficient. It
    does everything aside from check out a commit.
    If the version string is not specified, it uses git rev-parse HEAD.

* Incremental packing.
  * Pack as we go, to limit peak storage requirements.
    * Result: I see ~70-80GiB storage needed to build one month with 20
      parallel builds, which can be comfortably done in tmpfs on
      a big enough machine.

* Idempotency and purity (robustness).
  * Interrupting and rerunning should give the same result as if it had
    not been interrupted.
    * An improvement I could make here is to skip the build if the
      snapshot already exists.
  * Running multiple concurrently must not interfere with one another
    (other than 'safe' caches).
  * Git operations are done without creating separate worktrees; there
    is no shared state between builds- instead we use something like
    `GIT_WORK_TREE=. GIT_DIR=/the/repo/.git GIT_INDEX_FILE=./index git restore --worktree --staged --source=$COMMIT`.
    This is idempotent (it is not bothered by existing repositories),
    and it does an incremental checkout, which is fast, taking only
    milliseconds to move between commits. It has no shared state with
    any other repository.
  * Aiming for the same result out of the build system for a commit as
    if you had checked out that commit and built it for yourself.
  * Avoid embedding the compiler ident into the object files
    (-fno-ident). This is necessary because I built using a custom
    PGO/LTO build of clang-12; it produces the same output as a system
    clang-12 modulo the ident, but is substantially faster.

* CMakeCache safety.
  * It's conceivable that someone could make a change to the build
    system, which on a second run resulted in a different cache. To
    avoid this, we query if the build system needs to be rerun, and if
    it does, purge the build system cache. Build system outputs are
    still kept though and can continue to be used.
    * (This was done before but now is done in a more complete way by
      asking if build.ninja is stale).

* Safety in the face of objects and headers being renamed.
  * If a header or object is renamed or removed from the build system,
    the file generated in a previous build needs to be deleted. To that end
    we query the build system to ask which objects it would delete, compare
    with the filesystem, and delete the unneeded ones.

* Efficiency.
  * Given that this runs on a machine with a large number of CPUs, we
    want to keep those cores as busy as we can without overcommit
    inefficiency. To that end, we use a jobserver, which ensures that
    compilation jobs are matched with cores.
    * See also:
      * https://github.com/ninja-build/ninja/issues/1139
      * https://github.com/olsner/jobclient
  * The build for a single commit often ends up bottlenecked on
    individual compilation units, turning it into a serial process.
    To keep all the cores busy, we split the list of all commits into
    chunks of ~100ish each, and build many 100-commit chunks in
    parallel. They all run under the same jobserver which prevents OOM
    and avoids CPU overcommit inefficiency.

* Speed.
  * Avoid copying or updating files where not neccessary.
  * Only build the objects we need, and don't bother with linking the
    binaries.
    * This is achieved by inserting a 'manyclangs-targets' rule at the
      end of the build.ninja, containing all .o outputs and
      clang-resource-headers, which is necessary because the
      command line argument limit is hit if passing them directly as
      targets.
  * Concurrently build a single month.
    * The goal here was to reduce turnaround time of getting a single
      month built for testing/validation purposes.
  * Result: Building a pack for one month now takes in the region of
    3-4 hours wall time on a reasonable 64-core aarch64 machine.

* Document some of the rationale behind our cmake configuration.
* Various other fixes.

Note for any reviewers:

* A lot of stuff changed, sorry. I'm not expecting a super thorough
  review, my belief is that the proof is in the execution of this and
  validation.
* Very happy if you spot any issues though!

### CODE REVIEW

**Comment by peterwaller-arm** [comment]
This branch now reflects the state of the scripts used to build the packs in the manyclangs v0.9.0 release.

File: contrib/manyclangs-build-month [hunk 1]
```diff
@@ -24,64 +24,209 @@ list_month() {
...
+        export -f _manyclangs_build_parts _manyclangs_build_part _manyclangs_list_files0_to_pack
+        # A few extra slots, since we loose at least one to the build system and we
+        # don't expect every job to take 100% CPU.
+        JOBSLOP=5
+        command time -v \
+            jobserver -j$(($(nproc) + JOBSLOP)) \
+            xargs --max-procs=$NPARALLEL -n1 \
+            bash -c '_manyclangs_build_parts "$@"' -- "$YEAR" "$MONTH" \
+                <<<$(seq -f '%02.0f' 00 $(( NPARTS - 1 ))) \
+                | progress_bar "$COMMIT_COUNT" \
```
Comment (veselink1) [inline contrib/manyclangs-build-month:201]: Ah, nice!

File: contrib/manyclangs-build-month [hunk 2]
```diff
@@ -24,64 +24,209 @@ list_month() {
...
+
+    commits_for_month "$YEAR" "$MONTH"  > manyclangs-commit-list.txt
+    apply_limit < manyclangs-commit-list.txt | apply_subset > manyclangs-commit-list.txt.subset
+    mv manyclangs-commit-list.txt.subset manyclangs-commit-list.txt
+    COMMIT_COUNT=$(wc -l manyclangs-commit-list.txt | awk '{print $1}')
+
+    NPARALLEL=${MANYCLANGS_NPARALLEL-5} # Number of parts to run in parallel.
+
+    if [[ "$NPARALLEL" == 1 ]] || command -v ninja-jobclient &> /dev/null
+    then
```
Comment (veselink1) [inline contrib/manyclangs-build-month:188]: Why take this code path if `NPARALLEL=1`? Shouldn't control flow move to the `else`, with serial compilation?

File: contrib/manyclangs-build-month [hunk 3]
```diff
@@ -24,64 +24,209 @@ list_month() {
...
+        NPARTS=${MANYCLANGS_NPARTS-20} # Divide work into small equal-ish-sized parts.
+        split --number "l/$NPARTS" --numeric-suffixes manyclangs-commit-list.txt{,.}
+
+        export -f _manyclangs_build_parts _manyclangs_build_part _manyclangs_list_files0_to_pack
+        # A few extra slots, since we loose at least one to the build system and we
+        # don't expect every job to take 100% CPU.
+        JOBSLOP=5
+        command time -v \
+            jobserver -j$(($(nproc) + JOBSLOP)) \
```
Comment (veselink1) [inline contrib/manyclangs-build-month:197]: Which package installs `jobserver`? Can we mention that somewhere?

**Review by veselink1 — APPROVED** [review]
Awesome work, thanks!

File: contrib/manyclangs-build-month (hunk 3 above)
Comment (peterwaller-arm) [reply contrib/manyclangs-build-month:197]: Sadly there is no package, but it is available at https://github.com/olsner/jobclient

I agree this needs mentioning.

**Pull request merged**